### PR TITLE
[InstCombine] Fold `icmp eq/ne min|max(X, Y), Z`

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch further improves the simplification of pattern `icmp eq/ne min|max(X, Y), Z` as discussed in [D156238](https://reviews.llvm.org/D156238).
When `X < Z`:
`min(X, Y) == Z -> false`
`min(X, Y) != Z -> true`
`max(X, Y) == Z -> Y == Z`
`max(Y, Z) != Z -> Y != Z`
When `X > Z`:
`max(X, Y) == Z -> false`
`max(X, Y) != Z -> true`
`min(X, Y) == Z -> Y == Z`
`min(Y, Z) != Z -> Y != Z`

Alive2:  https://alive2.llvm.org/ce/z/evkmaq

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 1]
```diff
@@ -5022,6 +5022,12 @@ InstCombinerImpl::foldICmpWithMinMaxImpl(Instruction &I,
     std::swap(CmpXZ, CmpYZ);
   }
 
+  auto FoldIntoCmpYZ = [&]() -> Instruction * {
+    if (CmpYZ.has_value())
+      return replaceInstUsesWith(I, ConstantInt::getBool(I.getType(), *CmpYZ));
+    return ICmpInst::Create(Instruction::ICmp, Pred, Y, Z);
+  };
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:5029]: This lambda is only used once.
Imo its clearer to just leave this logic inlined at its callsite.

Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:5029]: It is also used in
https://github.com/llvm/llvm-project/blob/07151f0241d3f893cb36eb2dbc395d4098f74a87/llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp#L5079
and
https://github.com/llvm/llvm-project/blob/07151f0241d3f893cb36eb2dbc395d4098f74a87/llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp#L5088

File: llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp [hunk 2]
```diff
@@ -5038,12 +5044,32 @@ InstCombinerImpl::foldICmpWithMinMaxImpl(Instruction &I,
...
-    // min(X, Y) != Z X <= Y && Y != Z
-    // max(X, Y) != Z X >= Y && Y != Z
+    // Otherwise (X != Z):
+    ICmpInst::Predicate NewPred = MinMax->getPredicate();
+    auto NewCmpXZ = IsCondKnownTrue(simplifyICmpInst(NewPred, X, Z, Q));
+    if (!NewCmpXZ.has_value()) {
+      std::swap(X, Y);
+      std::swap(CmpXZ, CmpYZ);
+      NewCmpXZ = IsCondKnownTrue(simplifyICmpInst(NewPred, X, Z, Q));
+    }
```
Comment (goldsteinn) [inline llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:5054]: This is also applicable to other predicates no?
I.e
`max(X, Y) u<= Z`
If you have `MinMax`(`UGT`) known true -> true. Known false -> `Y u<= Z`.

Comment (dtcxzyw) [reply llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:5054]: It has been addressed by [D156238](https://reviews.llvm.org/D156238).

https://github.com/llvm/llvm-project/blob/07151f0241d3f893cb36eb2dbc395d4098f74a87/llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp#L5064-L5097

**Comment by goldsteinn** [comment]
LGTM.

**Comment by chapuni** [comment]
Seems this broke bootstrapping in some builds.
eg. https://lab.llvm.org/buildbot/#/builders/198/builds/5082

**Comment by dtcxzyw** [comment]
@chapuni @vitalybuka Thank you for reverting this! I am investigating the regression.

**Comment by dtcxzyw** [comment]
Relanded as 416e891d6609d919211db7ab14149970ec49b0f6

**Pull request merged**